### PR TITLE
Bug fix in MP3 and FLAC compute length on TTSDataset

### PR DESCRIPTION
**Pull request by Edresson**

### CODE REVIEW

File: TTS/tts/datasets/dataset.py [hunk 1]
```diff
@@ -42,6 +44,19 @@ def string2filename(string):
...
+    if audiopath[-4:] == ".mp3":
+        audio_info = MP3(audiopath).info
+        return int(audio_info.length * audio_info.sample_rate)
+    elif audiopath[-4:] == ".wav" or audiopath[-5:] == ".flac":
+        compress_factor = 8
+        bitrate = 16 # assuming 16bit audio
+        return int(os.path.getsize(audiopath) / bitrate * compress_factor)
+    else:
+        audio_format = audiopath.split(".")[-1]
+        raise RuntimeError(f"The audio format {audio_format} is not supported, please convert the audio files for mp3, flac or wav format!")
```
Comment (akx) [inline TTS/tts/datasets/dataset.py:57]: Might be better to just compute the extension once (and also make the comparison case-insensitive):

```suggestion
    extension = audiopath.rpartition(".")[-1].lower()
    if extension == "mp3":
        audio_info = MP3(audiopath).info
        return int(audio_info.length * audio_info.sample_rate)
    if extension in ("wav", "flac"):
        compress_factor = 8
        bitrate = 16 # assuming 16bit audio
        return int(os.path.getsize(audiopath) / bitrate * compress_factor)
    raise RuntimeError(f"The audio format {extension} is not supported, please convert the audio files for mp3, flac or wav format!")
```

File: TTS/tts/datasets/dataset.py [hunk 2]
```diff
@@ -42,6 +44,19 @@ def string2filename(string):
     return filename
 
 
+def get_audio_size(audiopath):
+    if audiopath[-4:] == ".mp3":
+        audio_info = MP3(audiopath).info
+        return int(audio_info.length * audio_info.sample_rate)
+    elif audiopath[-4:] == ".wav" or audiopath[-5:] == ".flac":
```
Comment (akx) [inline TTS/tts/datasets/dataset.py:51]: This won't work correctly for FLACs; they're compressed data.

```
-rw-r--r--@  1 akx  staff  18658921 Mar 26  2023 vocals.flac
-rw-r--r--   1 akx  staff  30728372 Mar 26  2023 vocals.wav
```

IOW, the compression ratio here is ~60% of the original.

Comment (Edresson) [reply TTS/tts/datasets/dataset.py]: Indeed, good catch in fact it is dynamic. I have got ~51% for other audio.

Comment (Edresson) [reply TTS/tts/datasets/dataset.py]: mutagen library supports flac and wav. I have used it all the audio formats. In that way, we can get the length in a more precise way.

**Review by akx — COMMENTED** [review]
Nice! I guess a test that checks that each of the 3 ostensibly supported formats actually works (both `get_audio_size` and loading-wise) would be useful...

**Comment by Edresson** [comment]
> Nice! I guess a test that checks that each of the 3 ostensibly supported formats actually works (both `get_audio_size` and loading-wise) would be useful...

Yeah, I added a unit test that loads with all supported audio formats. And also `get_audio_size` function.

File: TTS/tts/datasets/dataset.py [hunk 3]
```diff
@@ -42,6 +46,20 @@ def string2filename(string):
     return filename
 
 
+def get_audio_size(audiopath):
+    extension = audiopath.rpartition(".")[-1].lower()
+    if extension == "mp3":
+        audio_info = MP3(audiopath).info
```
Comment (WeberJulian) [inline TTS/tts/datasets/dataset.py:52]: Is there a reason not to use `mutagen.File`?

**Review by WeberJulian — APPROVED** [review]
LGTM

File: TTS/tts/datasets/dataset.py (hunk 3 above)
Comment (Edresson) [reply TTS/tts/datasets/dataset.py]: Indeed, it is better. I didn't found mutagen.File when I search for it

**Review by WeberJulian — APPROVED** [review]

**Comment by Edresson** [comment]
Looks like the Trainer PR https://github.com/coqui-ai/Trainer/pull/131 brokes the aux tests

**Pull request merged**